rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,423 @@
+import tensorflow as tf
+from typing import Text, List, Dict, Any, Union, Optional, Tuple
+import tensorflow_addons as tfa
+
+from rasa.utils.tensorflow.model_data import FeatureSignature
+from rasa.utils.tensorflow.constants import (
+    REGULARIZATION_CONSTANT,
+    WEIGHT_SPARSITY,
+    NUM_TRANSFORMER_LAYERS,
+    TRANSFORMER_SIZE,
+    NUM_HEADS,
+    UNIDIRECTIONAL_ENCODER,
+    KEY_RELATIVE_ATTENTION,
+    VALUE_RELATIVE_ATTENTION,
+    MAX_RELATIVE_POSITION,
+    HIDDEN_LAYERS_SIZES,
+    DROP_RATE,
+    SPARSE_INPUT_DROPOUT,
+    DENSE_INPUT_DROPOUT,
+    DENSE_DIMENSION,
+    CONCAT_DIMENSION,
+    DROP_RATE_ATTENTION,
+    SEQUENCE,
+    SENTENCE,
+)
+from rasa.utils.tensorflow import layers
+from rasa.shared.nlu.constants import TEXT
+
+# TODO: use this? it's in layers.py
+tfa.options.TF_ADDONS_PY_OPS = True
+
+
+class ConcatenateSparseDenseFeatures(tf.keras.layers.Layer):
+    # TODO add docstring
+    def __init__(
+        self,
+        attribute: Text,
+        feature_type: Text,
+        data_signature: List[FeatureSignature],
+        dropout_rate: float,
+        sparse_dropout: bool,
+        dense_dropout: bool,
+        dense_concat_dimension: int,
+        sparse_to_dense_kw: Dict[Text, Any] = {},
+    ) -> None:
+        super().__init__(
+            name=f"concatenate_sparse_dense_features_{attribute}_{feature_type}"
+        )
+        self.have_sparse_features = any(
+            [signature.is_sparse for signature in data_signature]
+        )
+        self.have_dense_features = any(
+            [not signature.is_sparse for signature in data_signature]
+        )
+
+        all_sparse_units = sum(
+            [
+                dense_concat_dimension
+                for signature in data_signature
+                if signature.is_sparse
+            ]
+        )
+        all_dense_units = sum(
+            [signature.units for signature in data_signature if not signature.is_sparse]
+        )
+        self.output_units = all_sparse_units + all_dense_units
```

Comment:
could you please add comment, explaining this logic

Reply:
Sure. As I was rewriting this code a lot, I wanted to leave writing comments for later, but maybe now is the best time...